Add Config to Disable Hosts File Modification

tricktron opened the pull request:
## Description

This change introduces a new configuration setting `modify-hosts-file` that allows users to disable CRC's automatic modification of the /etc/hosts file. The setting defaults to true to maintain backward compatibility with existing installations.

When set to false, CRC will skip adding OpenShift hostnames to the system's hosts file during cluster startup. This is particularly useful for systems like NixOS where the hosts file is managed declaratively and cannot be modified by applications, or for users who prefer to manage DNS resolution through other means.

The implementation adds the configuration option to all platform-specific DNS modules (Linux, macOS, Windows) and includes the setting in the service configuration passed to DNS setup routines.

<!--
Thank you for your pull request (PR)!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes.
-->

<!--
Describe in plain English what you solved and how. For instance, _Added `start` command to CRC so the user can create a VM and set up a single-node OpenShift cluster on it with one command. It requires blablabla..._
-->

## Type of change
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change
- [ ] Chore (non-breaking change which doesn't affect codebase;
  test, version modification, documentation, etc.)

## Proposed changes
<!--
List main as well as consequential changes you introduced or had to introduce.

1. Add `start` command.
2. Add `setup` as prerequisite to `start`.
3. ...
-->

## Testing

I tested this e2e on my NixOS machine by manually setting the needed host file entries: `api.crc.testing, canary-openshift-ingress-canary.apps-crc.testing, ...` and with the new config  `modify-hosts-file = false`.
<!--
What is the _bottom-line_ functionality that needs testing? Describe in pseudo-code or in English. Use verifiable statements that tie your changes to existing functionality.

1. `start` succeeds first time after `setup` succeeded
2. stdout contains ... if `start` succeeded
3. stderr contains ... after `start` failed
4. `status` returns ... if `start` succeeded
5. `status` returns ... if `start` failed
6. `start` fails after `start` succeeded or after `status` says CRC is `Running`
7. ...
-->

## Contribution Checklist
- [x] I Keep It Small and Simple: The smaller the PR is, the easier it is to review and have it merged
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Which platform have you tested the code changes on? <!-- Only put an `x` in applicable platforms -->
    - [x] Linux
    - [ ] Windows
    - [ ] MacOS

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a user-configurable setting to control whether CRC updates the system hosts file during start and via API endpoints (default: enabled). When disabled, CRC skips modifying hosts on Linux, macOS, and Windows and logs the skip.

* **Documentation / Messaging**
  * Improved post-start DNS error guidance advising verification of system DNS/hosts when hosts modification is disabled.

* **Tests**
  * Extended tests to cover the new setting's default and provided values and API behavior.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

praveenkumar commented:
/ok-to-test

tricktron commented:
/retest-required

tricktron commented:
@praveenkumar Thanks for running the tests. I see many timeout failures unrelated to my change. Is this normal and can I rerun the tests or do only maintainers have the right to rerun tests?

praveenkumar commented:
/retest

gbraad commented on pkg/crc/services/dns/dns_darwin.go:
```diff
@@ -33,8 +33,10 @@ type resolverFileValues struct {
+	if serviceConfig.ModifyHostsFile {
+		if err := addOpenShiftHosts(serviceConfig); err != nil {
+			return err
+		}
```
```diff 
- }
+ } else {
+ 	logging.Infof("Skipping hosts file modification")
+ }
```

gbraad replied:
... this of course applies to all the platforms (also `dns_linux.go` and `dns_windows.go`).

tricktron replied:
@gbraad Thank you for your time and the review. I applied your suggestions and also the other two nitpicks by coderabbitai.

gbraad reviewed (changes requested):
Suggest minor changes to ensure this is recorded as part of log

tricktron commented:
@gbraad Kind ping to have another look :smiley:.

tricktron commented:
@gbraad Another friendly ping to have look 😃.

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -694,7 +699,7 @@ func createHost(machineConfig config.MachineConfig, preset crcPreset.Preset) err
 			return errors.Wrap(err, "Error generating new kubeadmin password")
 		}
-		if err = os.WriteFile(constants.GetDeveloperPasswordPath(), []byte(constants.DefaultDeveloperPassword), 0600); err != nil {
+		if err = os.WriteFile(constants.GetDeveloperPasswordPath(), []byte(constants.DefaultDeveloperPassword), 0o600); err != nil {
```
This change shouldn't be part of commit.

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -748,7 +753,7 @@ func enableEmergencyLogin(sshRunner *crcssh.Runner) error {
 		b[i] = charset[rand.Intn(len(charset))] //nolint
 	}
-	if err := os.WriteFile(constants.PasswdFilePath, b, 0600); err != nil {
+	if err := os.WriteFile(constants.PasswdFilePath, b, 0o600); err != nil {
```
same as above

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -775,7 +780,7 @@ func updateSSHKeyPair(sshRunner *crcssh.Runner) error {
 
 	logging.Info("Updating authorized keys...")
-	err = sshRunner.CopyData(publicKey, "/home/core/.ssh/authorized_keys", 0644)
+	err = sshRunner.CopyData(publicKey, "/home/core/.ssh/authorized_keys", 0o644)
```
same

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -874,10 +879,10 @@ func startMicroshift(ctx context.Context, sshRunner *crcssh.Runner, ocConfig oc.
 		return err
 	}
-	if err := sshRunner.CopyFileFromVM(fmt.Sprintf("/var/lib/microshift/resources/kubeadmin/api%s/kubeconfig", constants.ClusterDomain), constants.KubeconfigFilePath, 0600); err != nil {
+	if err := sshRunner.CopyFileFromVM(fmt.Sprintf("/var/lib/microshift/resources/kubeadmin/api%s/kubeconfig", constants.ClusterDomain), constants.KubeconfigFilePath, 0o600); err != nil {
```
same

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -874,10 +879,10 @@ func startMicroshift(ctx context.Context, sshRunner *crcssh.Runner, ocConfig oc.
 		return err
 	}
-	if err := sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0644); err != nil {
+	if err := sshRunner.CopyFile(constants.KubeconfigFilePath, "/opt/kubeconfig", 0o644); err != nil {
```
same

praveenkumar commented on pkg/crc/machine/start.go:
```diff
@@ -895,5 +900,5 @@ func ensurePullSecretPresentInVM(sshRunner *crcssh.Runner, pullSec cluster.PullS
 		return err
 	}
-	return sshRunner.CopyDataPrivileged([]byte(content), "/etc/crio/openshift-pull-secret", 0600)
+	return sshRunner.CopyDataPrivileged([]byte(content), "/etc/crio/openshift-pull-secret", 0o600)
```
same

praveenkumar reviewed (commented):
Changes regarding file permission change should be in separate commit.

tricktron committed:
Use explicit octal prefix for file permissions

Convert legacy octal notation (0600, 0644) to modern Go style with
explicit 0o prefix (0o600, 0o644) for improved code clarity.

tricktron committed:
Add config to disable hosts file modification

This change introduces a new configuration setting 'modify-hosts-file'
that allows users to disable CRC's automatic modification of the
/etc/hosts file. The setting defaults to true to maintain backward
compatibility with existing installations.

tricktron commented:
> Changes regarding file permission change should be in separate commit.

@praveenkumar I created a separate commit for it :+1:.

praveenkumar commented:
@tricktron https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns.go#L35-L37 have you also take account to this one because it run https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns_linux.go => https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns.go#L207 which means this entry still added to /etc/hosts.

tricktron commented:
> @tricktron https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns.go#L35-L37 have you also take account to this one because it run https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns_linux.go => https://github.com/crc-org/crc/blob/main/pkg/crc/services/dns/dns.go#L207 which means this entry still added to /etc/hosts.

@praveenkumar Yes, as far as I can see I have guarded all calls to the `addOpenShiftHosts` function behind the new `ModifyHostsFile` config flag.

praveenkumar commented:
@tricktron I got some time today to test the PR and I think https://github.com/crc-org/crc/pull/4938#issuecomment-3632510911 still make sense because I still see entry on /etc/hosts file
```
$ ./crc config view
- memory                                : 13000
- modify-hosts-file                     : false

$ cat /etc/hosts
# Loopback entries; do not change.
# For historical reasons, localhost precedes localhost.localdomain:
127.0.0.1        localhost localhost.localdomain localhost4 localhost4.localdomain4
::1              localhost localhost.localdomain localhost6 localhost6.localdomain6

# Added by CRC
127.0.0.1        oauth-openshift.apps-crc.testing console-openshift-console.apps-crc.testing downloads-openshift-console.apps-crc.testing default-route-openshift-image-registry.apps-crc.testing
# End of CRC section
```

tricktron committed:
Gate daemon host file endpoints

Guard host file endpoints behind the modify-hosts-file setting so the
daemon does not touch hosts entries when the feature is disabled. Adds
tests to cover the disabled and enabled paths.

tricktron commented:
@praveenkumar I also gated the daemon calls to the `/etc/hosts` file and added some tests for it. Can you test again?

tricktron commented:
@praveenkumar Friendly ping to have another look at this.

praveenkumar commented:
> @praveenkumar Friendly ping to have another look at this.

@tricktron I will take another look on Tuesday and hopefully we can merge it.

praveenkumar commented:
/cherry-pick crc-ng
